File: tool/GraceRead_L1_2008_03_20/lib/gps2utc.c
/* Purpose:  to return the UTC time in secons past J2000
             given gps time. The time is given in gps seconds past J2000. */

  /* Input:  
     real gps_time        gps time given in seconds past J2000
     table table_leap     table containing leapseconds (table_leap.x[]), 
                          the time of the leapseconds (table_leap.y[]),
                          and the number of entries(table_leap.size)
   
     Output:
     real (return value)  utc time given in seconds past J2000  */

#include <stdio.h>
#include "TimeLib.h" 

real gps2utc(real gps_time, table table_leap)
{
  static char SccsId[] = "@(#) gps2utc.c       1.1 01/08/01";

  /* Local variables: */ 
  real utc_t1;            /* utc time in seconds past J2000 */ 
  real b4lp;              /* time of leapsecond before input time */
  int lpsec;              /* number of gps leapseconds */

/* Call gpslpsec twice to avoid being off by one second when */
/* the utc time is less than gpslpsec before a leap second   */

  gpslpsec(gps_time,&lpsec,&b4lp,table_leap);
  utc_t1 = gps_time - lpsec;

  /* Without the following fix, the following would happen: 

  1-JUL-1993 00:00:10 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
  1-JUL-1993 00:00:09 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
  1-JUL-1993 00:00:08 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
  1-JUL-1993 00:00:07 GPS TIME ---> 30-JUN-1993 23:59:59 UTC TIME

  The fix allows the following conversion: 

  1-JUL-1993 00:00:10 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
  1-JUL-1993 00:00:09 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
  1-JUL-1993 00:00:08 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
  1-JUL-1993 00:00:07 GPS TIME ---> 30-JUN-1993 23:59:59 UTC TIME */

  if( utc_t1 == b4lp )
  {
    return (gps_time - lpsec);
  }

  gpslpsec(utc_t1,&lpsec,&b4lp,table_leap);

  return(gps_time-lpsec);
}

File: lib/lib_timesys/gps2utc.f
*
*  procedure     GPS2UTC
*
      double precision Function GPS2UTC(GPS)
*********1*********2*********3*********4*********5*********6*********7**
*      
*                           Purpose
*
*  Return the UTC time in secons past J2000 given gps time. The time is
*  given in gps seconds past J2000.
*
*                           Input_Output Auguments
*
*  Name     Type    I/O     Descriptin
*  ----     ----    ---     --------------------------------------------
*  GPS      REAL*8  I       gps time given in seconds past J2000
*  UTC      REAL*8  O       utc time given in seconds past J2000 
*      
*  History:
*
*  Vesion 1.0
*    
*  Time         Author      Description
*  ----         ------      --------------------------------------------
*  07.06.01     S.J.Zhang   build this subroutine    
*
*********1*********2*********3*********4*********5*********6*********7**
c
c     Declaration_of_Varialbes
c
      IMPLICIT NONE
C
C     Declaration_of_Constants
C
      INTEGER       MAX_LEAP
      PARAMETER(    MAX_LEAP = 20 )
c
c     Declaration_of_Input_Output_Varialbles
c
      REAL*8        GPS, UTC
C
C     Declaration_of_External_Functions
C
      REAL*8        JD2SEC
      REAL*8        gpslpsec
C
C     Declaration_of_Local_Variables
C
      INTEGER       NLEAP
      INTEGER       I, J
      REAL*8        LEAP(MAX_LEAP,2)
      REAL*8        UTC_t1
      REAL*8        UTC_leapsec
c
c     Call gpslpsec twice to avoid being off by one second when 
c     the utc time is less than gpslpsec before a leap second 
c
      call leapsec(nleap, leap)
c
c     Get the leap second time before the GPS time
c
      do i=1, nleap
        if(GPS.ge.leap(i,1))then
          UTC_leapsec = leap(i,1)
        endif 
      enddo
c
      UTC_T1 = GPS - gpslpsec(GPS)
c
c     Without the following fix, the following would happen: 
******
c     1-JUL-1993 00:00:10 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
c     1-JUL-1993 00:00:09 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
c     1-JUL-1993 00:00:08 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
c     1-JUL-1993 00:00:07 GPS TIME ---> 30-JUN-1993 23:59:59 UTC TIME
******
c     The fix allows the following conversion: 
******
c     1-JUL-1993 00:00:10 GPS TIME --->  1-JUL-1993 00:00:01 UTC TIME
c     1-JUL-1993 00:00:09 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
c     1-JUL-1993 00:00:08 GPS TIME --->  1-JUL-1993 00:00:00 UTC TIME
c     1-JUL-1993 00:00:07 GPS TIME ---> 30-JUN-1993 23:59:59 UTC TIME 
c
      UTC = GPS - gpslpsec(UTC_T1)
c
      GPS2UTC = UTC
c
      return
c
      end

